internal/querynodev2/pipeline/embedding_node.go

```diff
@@ -0,0 +1,249 @@
+		zap.Int("insertRowNum", len(pks)),
+		zap.Uint64("timestampMin", msg.BeginTimestamp),
+		zap.Uint64("timestampMax", msg.EndTimestamp))
+}
+
+func (eNode *embeddingNode) bm25Embedding(runner function.FunctionRunner, msg *msgstream.InsertMsg, stats map[int64]*storage.BM25Stats) error {
+	functionSchema := runner.GetSchema()
+	inputFieldID := functionSchema.GetInputFieldIds()[0]
+	outputFieldID := functionSchema.GetOutputFieldIds()[0]
+	outputField := runner.GetOutputFields()[0]
+
+	// TODO REMOVE CODE FOR TEST
+	// REMOVE invalid vector field data used as placeholder
+	msg.FieldsData = RemoveFieldData(msg.GetFieldsData(), outputFieldID)
+
+	data, err := GetEmbeddingFieldData(msg.GetFieldsData(), inputFieldID)
+	if data == nil || err != nil {
+		return merr.WrapErrFieldNotFound(fmt.Sprint(inputFieldID))
+	}
+
+	output, err := runner.BatchRun(data)
+	if err != nil {
+		return err
+	}
+
+	if _, ok := stats[outputFieldID]; !ok {
+		stats[outputFieldID] = storage.NewBM25Stats()
+	}
+
+	sparseMaps, ok := output[0].([]map[uint32]float32)
+	if !ok {
+		return fmt.Errorf("bm25 runner return unknown type output")
+	}
+	stats[outputFieldID].Append(sparseMaps...)
+	dim := 0
+	sparseVector := lo.Map(sparseMaps, func(sparseMap map[uint32]float32, _ int) []byte {
+		if len(sparseMap) > dim {
+			dim = len(sparseMap)
+		}
+		return typeutil.CreateAndSortSparseFloatRow(sparseMap)
+	})
+	msg.FieldsData = append(msg.FieldsData, delegator.BuildSparseFieldData(outputField, int64(dim), sparseVector))
+	return nil
+}
+
+func (eNode *embeddingNode) embedding(msg *msgstream.InsertMsg, stats map[int64]*storage.BM25Stats) error {
+	for _, functionRunner := range eNode.functionRunners {
+		functionSchema := functionRunner.GetSchema()
+		switch functionSchema.GetType() {
+		case schemapb.FunctionType_BM25:
+			err := eNode.bm25Embedding(functionRunner, msg, stats)
+			if err != nil {
+				return err
+			}
+		default:
+			log.Warn("pipeline embedding with unkonwn function type", zap.Any("type", functionSchema.GetType()))
+			return fmt.Errorf("Unknown function type")
+		}
+	}
+
+	return nil
+}
+
+func (eNode *embeddingNode) Operate(in Msg) Msg {
+	nodeMsg := in.(*insertNodeMsg)
+	nodeMsg.insertDatas = make(map[int64]*delegator.InsertData)
+
+	collection := eNode.manager.Collection.Get(eNode.collectionID)
+	if collection == nil {
+		log.Error("insertNode with collection not exist", zap.Int64("collection", eNode.collectionID))
+		panic("insertNode with collection not exist")
+	}
+
+	for _, msg := range nodeMsg.insertMsgs {
+		eNode.addInsertData(nodeMsg.insertDatas, msg, collection)
+	}
+
+	return nodeMsg
+}
+
+func GetEmbeddingFieldData(datas []*schemapb.FieldData, fieldID int64) ([]string, error) {
+	for _, data := range datas {
+		if data.GetFieldId() == fieldID {
+			return data.GetScalars().GetStringData().GetData(), nil
+		}
+	}
+	return nil, fmt.Errorf("field%d not found", fieldID)
+}
+
+func GetSparseVectorDim(data [][]byte) int64 {
```

Comment:
function is unused. If necessary, place it in file `SparseFloatRowDim` is defined